pkg/github/github.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package github
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/google/go-github/v29/github"
+	"github.com/pkg/errors"
+
+	"k8s.io/release/pkg/git"
+	"k8s.io/release/pkg/util"
+)
+
+// GitHub is a wrapper around GitHub related functionality
+type GitHub struct {
+	client Client
+}
+
+type githubClient struct {
+	*github.Client
+}
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -generate
+//counterfeiter:generate . Client
+type Client interface {
+	ListTags(
+		context.Context, string, string, *github.ListOptions,
+	) ([]*github.RepositoryTag, *github.Response, error)
+}
+
+// New creates a new default GitHub client
+func New() *GitHub {
+	return &GitHub{&githubClient{github.NewClient(nil)}}
+}
+
+func (g *githubClient) ListTags(
+	ctx context.Context, owner, repo string, opt *github.ListOptions,
+) ([]*github.RepositoryTag, *github.Response, error) {
+	return g.Client.Repositories.ListTags(ctx, owner, repo, opt)
+}
+
+// SetClient can be used to manually set the internal GitHub client
+func (g *GitHub) SetClient(client Client) {
+	g.client = client
+}
+
+// TagsPerBranch is an abstraction over a simple branch to latest tag association
+type TagsPerBranch map[string]string
+
+// LatestGitHubTagsPerBranch returns the latest GitHub available tag for each
+// branch
+func (g *GitHub) LatestGitHubTagsPerBranch() (TagsPerBranch, error) {
+	allTags, _, err := g.client.ListTags(
+		context.Background(), git.DefaultGithubOrg, git.DefaultGithubRepo, nil,
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to retrieve GitHub releases")
+	}
+
+	releases := make(TagsPerBranch)
+	for _, t := range allTags {
+		tag := t.GetName()
+
+		// Alpha releases are only available on the master branch
+		if strings.Contains(tag, "alpha") {
+			releases.addIfNotExisting(git.Master, tag)
+			continue
+		}
+
+		// This should always succeed
+		semverTag, err := util.TagStringToSemver(tag)
+		if err != nil {
+			return nil, errors.Wrapf(err, "tag %s is not a vaild semver", tag)
+		}
+
+		// Latest vx.x.0 release are on both master and release branch
+		if len(semverTag.Pre) == 0 {
+			releases.addIfNotExisting(git.Master, tag)
+		}
```

Comment:
I assume that this is still correct since we *do not* cut an `v1.19.0-alpha.x`  together with `v1.18.0`, right?

Reply:
@saschagrunert are you saying that tags like `v1.17.0` or `v1.18.0-beta.0` exist on `master`?

Reply:
Hehe this is tricky. Technically beta, rc and final tags are on the release branch, that’s true.

But if we want to get the latest tag for a branch in terms of changelog generation, we might want to treat final major versions (1.x.0) as release for the master branch. (I think this is incredible confusing, honestly.)

Reply:
Ah I see! If that is what we are doing here, could we change the name / description of this function to indicate that? Since this is in the `github` pkg and not a pkg just for release notes I want to make sure we are explicit :)

Reply:
Yeah I’ll note that there, totally makes sense. 🙂

Reply:
See https://github.com/kubernetes/release/pull/1136/files#r384955706